### PR TITLE
Many things aren't lists in python 3

### PR DESCRIPTION
**Pull request by dpoirier**
In Python 3, many things that used to return lists now return iterators.
Examples: map, filter, zip, dict.keys().

This becomes important because we can't do the same operations on iterators
as we can on lists - we can't index into them, take their length,
compare them to a list, treat them as a boolean, or iterate over them
more than once.

Additionally, dict.keys() doesn't necessarily return the keys in the same
order in Python 3 as it did in Python 2, so in tests it's safest to
sort the result and compare to a previously-sorted result.

### CODE REVIEW

File: EquiTrack/partners/tests/test_serializers.py [hunk 1]
```diff
@@ -108,7 +108,7 @@ def test_create_fail_country_programme_required_for_PCA(self):
         exception = context_manager.exception
 
         self.assertIsInstance(exception.detail, dict)
-        self.assertEqual(exception.detail.keys(), ['country_programme'])
+        self.assertEqual((exception.detail.keys()), ['country_programme'])
```
Comment (reinbach) [inline EquiTrack/partners/tests/test_serializers.py]: Looks like we're missing `list`

File: EquiTrack/partners/tests/test_serializers.py [hunk 2]
```diff
@@ -703,7 +703,7 @@ def test_retrieve(self):
         serializer = PartnerOrganizationDetailSerializer(instance=self.partner)
 
         data = serializer.data
-        self.assertItemsEqual(data.keys(), [
+        self.assertItemsEqual(sorted(data.keys()), [
```
Comment (reinbach) [inline EquiTrack/partners/tests/test_serializers.py]: Don't think we need `sorted` here as we're using `assertItemsEqual` which I think we actually may want to change to `six.assertCountEqual` or change `assertItemsEqual` to `assertEqual` and leave `sorted` call in

File: EquiTrack/partners/tests/test_serializers.py [hunk 3]
```diff
@@ -713,12 +713,12 @@ def test_retrieve(self):
             'vendor_number', 'vision_synced'
         ])
 
-        self.assertItemsEqual(data['planned_engagement'].keys(), [
+        self.assertItemsEqual(sorted(data['planned_engagement'].keys()), [
```
Comment (reinbach) [inline EquiTrack/partners/tests/test_serializers.py]: We have `assertItemsEqual` and `sorted` together

File: EquiTrack/partners/tests/test_serializers.py [hunk 4]
```diff
@@ -713,12 +713,12 @@ def test_retrieve(self):
...
             'id', 'scheduled_audit', 'special_audit', 'spot_check_follow_up_q1', 'spot_check_follow_up_q2',
             'spot_check_follow_up_q3', 'spot_check_follow_up_q4', 'spot_check_mr',
             'total_spot_check_follow_up_required', 'spot_check_required', 'required_audit'
         ])
 
         self.assertEquals(len(data['staff_members']), 1)
-        self.assertItemsEqual(data['staff_members'][0].keys(), [
+        self.assertItemsEqual(sorted(data['staff_members'][0].keys()), [
```
Comment (reinbach) [inline EquiTrack/partners/tests/test_serializers.py]: We have `assertItemsEqual` and `sorted` together

File: EquiTrack/partners/tests/test_views.py [hunk 5]
```diff
@@ -659,15 +659,16 @@ def test_api_partners_retrieve(self):
...
-        self.assertEqual(['outstanding_findings', 'audits', 'programmatic_visits', 'spot_checks'],
-                         response.data['hact_values'].keys())
+        self.assertEqual(['programme_visits', 'spot_checks'],
+                         sorted(response.data['hact_min_requirements'].keys()))
+        self.assertEqual(['audits', 'outstanding_findings', 'programmatic_visits', 'spot_checks'],
+                         sorted(response.data['hact_values'].keys()))
         self.assertItemsEqual(
             ['completed', 'minimum_requirements'],
-            response.data['hact_values']['audits'].keys()
+            sorted(response.data['hact_values']['audits'].keys())
```
Comment (reinbach) [inline EquiTrack/partners/tests/test_views.py]: We have `assertItemsEqual` and `sorted` together

File: EquiTrack/reports/tests/test_views.py [hunk 6]
```diff
@@ -67,7 +67,9 @@ def test_api_results_list(self):
         url = reverse('results-list')
         response = self.forced_auth_req('get', url)
         self.assertEqual(response.status_code, status.HTTP_200_OK)
-        self.assertEqual(int(response.data[0]["id"]), self.result1.id)
+        response_ids = [int(item['id']) for item in response.data]
+        result_ids = [self.result1.id, self.result2.id]
+        self.assertEqual(sorted(response_ids), sorted(result_ids))
```
Comment (reinbach) [inline EquiTrack/reports/tests/test_views.py]: may be easier/simpler to use `six.assertCountEqual`, then again maybe not :)

File: EquiTrack/t2f/helpers/cost_summary_calculator.py [hunk 7]
```diff
@@ -245,14 +245,16 @@ def get_by_day_grouping(self):
...
+        for date in sorted(mapping.keys()):
+            itinerary = mapping[date]
             dto = DSAdto(date, itinerary)
             over_60 = counter > 60
             dto.daily_rate = self.get_dsa_amount(dto.region, over_60)
             dsa_dto_list.append(dto)
             counter += 1
 
-        return sorted(dsa_dto_list, cmp=lambda x, y: cmp(x.date, y.date))
+        return dsa_dto_list
```
Comment (reinbach) [inline EquiTrack/t2f/helpers/cost_summary_calculator.py]: :+1:

File: EquiTrack/tpm/models.py [hunk 8]
```diff
@@ -129,17 +129,17 @@ def has_action_permission(self, user=None, action=None):
...
+            'tpm_activities': [a.get_mail_context() for a in activities],
+            'multiple_tpm_activities': activities.count() > 1,
             'object_url': self.get_object_url(),
-            'partners': ', '.join(set(map(lambda a: a.partner.name, self.tpm_activities.all()))),
-            'interventions': ', '.join(set(map(
-                lambda a: a.intervention.title,
-                self.tpm_activities.filter(intervention__isnull=False)
-            ))),
+            'partners': ', '.join(partner_names),
+            'interventions': ', '.join(interventions),
```
Comment (reinbach) [inline EquiTrack/tpm/models.py]: :+1:

File: EquiTrack/utils/permissions/metadata.py [hunk 9]
```diff
@@ -18,14 +20,10 @@ def _remove_read_only(self, field_info):
...
-        else:
-            filtered_fields = map(lambda f: f.field_name, serializer._writable_fields)
-
-        for field_name in info.keys():
-            if field_name not in filtered_fields:
-                del info[field_name]
+        method = serializer.context['request'].method
+        fields = serializer._readable_fields if method == 'GET' else serializer._writable_fields
+        field_names = [f.field_name for f in fields]
+        info = {k: v for k, v in six.iteritems(info) if k in field_names}
```
Comment (reinbach) [inline EquiTrack/utils/permissions/metadata.py]: :+1:

File: EquiTrack/partners/tests/test_serializers.py (hunk 1 above)
Comment (dpoirier) [reply EquiTrack/partners/tests/test_serializers.py]: Oops, will fix.

File: EquiTrack/partners/tests/test_serializers.py (hunk 2 above)
Comment (dpoirier) [reply EquiTrack/partners/tests/test_serializers.py]: Good point, will fix in all the places.

(Saving six.assertCountEqual for a later PR.)

**Comment by dpoirier** [comment]
@reinbach updated

File: EquiTrack/utils/permissions/serializers.py [hunk 10]
```diff
@@ -150,7 +150,9 @@ def _readable_fields(self):
             self.permissions
         )
 
-        allowed_fields_names = map(lambda p: p.target.split('.')[-1], allowed_permissions)
+        allowed_fields_names = [
+            p.target.split('.')[-1] for p in allowed_permissions
+        ]
         disallowed_fields_names = map(lambda p: p.target.split('.')[-1], disallowed_permissions)
```
Comment (ntrncic) [inline EquiTrack/utils/permissions/serializers.py]: why is this one skipped? it's used in 169.

Comment (dpoirier) [reply EquiTrack/utils/permissions/serializers.py]: I was thinking we were just iterating through it so it was okay, but it's in a loop so maybe it's safer to just use a list here too.  I've changed it.

**Pull request merged**